quantumnetworks/analysis/error.py

```diff
@@ -0,0 +1,120 @@
+"""
+Error Analysis
+"""
+import numpy as np
+
+from quantumnetworks.systems.base import SystemSolver
+from typing import Any, Dict
+from abc import abstractmethod, ABCMeta
+
+from quantumnetworks.systems.multimode import MultiModeSystem
+from quantumnetworks.utils.visualization import plot_full_evolution
+from tqdm import tqdm
+
+
+class SystemError(metaclass=ABCMeta):
+    def __init__(self, system: SystemSolver, params_error: Dict[str, Any]) -> None:
+        self.system = system
+        self.params_error = params_error
+        self.solves = None
+        self.reset_solves()
+
+    def reset_solves(self):
+        self.solves = {"original": None, "with_error": [], "std": None}
+
+    @abstractmethod
+    def calculate_error(self, method: str, *args, parse_output=lambda X: X, **kwargs):
+        pass
+
+    def run(
+        self, method: str, *args, parse_output=lambda X: X, num_samples=11, **kwargs
+    ):
+        self.solves["original"] = parse_output(
+            getattr(self.system, method)(*args, **kwargs)
+        )
+        self.calculate_error(
+            method, *args, parse_output=parse_output, num_samples=num_samples, **kwargs
+        )
+        solves_with_error = np.array(self.solves["with_error"])
+        self.solves["std"] = np.std(solves_with_error, axis=0)
+
+    def plot(self, ts, **kwargs):
+        fig, ax = plot_full_evolution(
+            self.solves["original"],
+            ts,
+            xs_min=(self.solves["original"] - self.solves["std"]),
+            xs_max=(self.solves["original"] + self.solves["std"]),
+            **kwargs
+        )
+        ax.legend()
+        fig.tight_layout()
+        return fig, ax
+
+
+class MultiModeError(SystemError):
+    def calculate_error(
+        self, method: str, *args, num_samples=11, parse_output=lambda X: X, **kwargs
+    ):
```

Comment:
Copying the description over from #4, but good to add a docstring here to clarify the general idea of what this method is trying to do.

> _General idea:_
> Let's say we measure parameter `a ±  δa`, `b ±  δb`, and `c ±  δc`. Then, to find the error in `f(a,b,c)`, we can sample the parameter values of `a*` in the gaussian distribution centered around `a` with standard deviation of `δa` (and similarly for `b*` and `c*`) and calculate `f(a*,b*,c*)` multiple times. Then, we can take the standard deviation of that set of `f(a*,b*,c*)`s to find `δf`.